File: algorithms/3_hard/N/03620_network_recovery_pathways/binary_search.cpp
typedef  ptrdiff_t isize;
typedef  pair<int, int> DegCntUsedCnt;

class Solution {
    static inline vector<vector<pair<int, int>>> calc_adj(span<const vector<int>> edges, vector<bool>& online) {
        auto res = vector<vector<pair<int, int>>> (online.size(), vector<pair<int, int>>());
        for (const auto& e : edges) {
            const int u = e[0];
            const int v = e[1];
            const int cost = e[2];
            assert(u < std::ssize(online) && v < std::ssize(online) && cost >= 0);
            if (online[u] && online[v]) {
                res[u].push_back({v, cost});
            }
        }
        return res;
    }

    static inline vector<DegCntUsedCnt> calc_deg(span<vector<pair<int, int>>> adj) {
        auto res = vector<DegCntUsedCnt>(adj.size(), {0, 0});
        for (const auto& a : adj) {
            for (const auto [v, _] : a) {
                res[v].first += 1;
            }
        }

        const isize len = std::ssize(adj);
        for (isize u = 1; u < len; u += 1) {
            if (res[u].first == 0) {
                for (const auto [v, _] : adj[u]) {
                    res[v].first -= 1;
                }
            }
        }

        return res;
    }

    static bool check(span<vector<pair<int, int>>> adj, deque<int>& q, span<int64_t> costs, span<DegCntUsedCnt> deg, int mid, int target, int64_t k) {
        for (auto& c : costs) { c = numeric_limits<int64_t>::max() / 2; }
        for (auto& d : deg) { d.second = 0; }
        q.clear();
        q.push_back(0);
        costs[0] = 0;

        while (!q.empty()) {
            const int node = q.front();
            q.pop_front();
            const int64_t total_cost = costs[node];

            for (const auto [v, cost] : adj[node]) {
                deg[v].second += 1;
                if (deg[v].first == deg[v].second) {
                    q.push_back(v);
                }
                if (cost >= mid) {
                    costs[v] = std::min(costs[v], total_cost + cost);
                }
            }
        }

        return costs[target] <= k;
    }

public:
    static int findMaxPathScore(vector<vector<int>>& edges, vector<bool>& online, long long k) {
        if (edges.empty()) { return -1; }

        int lo = numeric_limits<int>::max();;
        int hi = 0;
        for (const auto& e : edges) {
            lo = std::min(lo, e[2]);
            hi = std::max(hi, e[2]);
        }

        const int target = static_cast<int>(std::ssize(online)) - 1;
        auto q = deque<int>();
        auto costs = vector<int64_t>(online.size(), 0);
        auto adj = calc_adj(edges, online);
        auto deg = calc_deg(adj);

        if (!check(adj, q, costs, deg, lo, target, k)) {
            return -1;
        }

        while (lo <= hi) {
            const int mid = lo + (hi - lo) / 2;
            if (check(adj, q, costs, deg, mid, target, k)) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return hi;
    }
};

File: algorithms/3_hard/N/03620_network_recovery_pathways/binary_search.rs
use std::collections::VecDeque;
type DegCntUsedCnt = (i32, i32);

impl Solution {
    #[inline(always)]
    fn calc_adj(edges: Vec<Vec<i32>>, online: Vec<bool>) -> Vec<Vec<(i32, i32)>> {
        let mut res: Vec<Vec<(i32, i32)>> = vec![Vec::default(); online.len()];
        for e in edges {
            let (u, v, cost) = (e[0], e[1], e[2]);
            debug_assert!((u as usize) < online.len() && (v as usize) < online.len() && cost >= 0);
            if online[u as usize] && online[v as usize] {
                res[u as usize].push((v, cost));
            }
        }
        res
    }

    #[inline(always)]
    fn calc_deg(adj: &[Vec<(i32, i32)>]) -> Vec<DegCntUsedCnt> {
        let mut res: Vec<DegCntUsedCnt> = vec![(0, 0); adj.len()];
        for a in adj {
            for &(v, _) in a {
                res[v as usize].0 += 1;
            }
        }

        for (u, a) in adj.iter().enumerate().skip(1) {
            if res[u as usize].0 == 0 {
                for &(v, _) in a {
                    res[v as usize].0 -= 1;
                }
            }
        }

        res
    }

    fn check(adj: &[Vec<(i32, i32)>], q: &mut VecDeque<i32>, costs: &mut [i64], deg: &mut [DegCntUsedCnt], mid: i32, target: i32, k: i64) -> bool {
        costs.iter_mut().for_each(|x| *x = i64::MAX / 2);
        deg.iter_mut().for_each(|x| x.1 = 0);
        q.clear();
        q.push_back(0);
        costs[0] = 0;

        while let Some(node) = q.pop_front() {
            let total_cost = costs[node as usize];

            for &(v, cost) in adj[node as usize].iter() {
                deg[v as usize].1 += 1;
                if deg[v as usize].0 == deg[v as usize].1 {
                    q.push_back(v);
                }
                if cost >= mid {
                    costs[v as usize] = costs[v as usize].min(total_cost + cost as i64);
                }
            }
        }

        costs[target as usize] <= k
    }

    pub fn find_max_path_score(edges: Vec<Vec<i32>>, online: Vec<bool>, k: i64) -> i32 {
        if edges.is_empty() { return -1; }

        let mut lo = i32::MAX;
        let mut hi = 0_i32;
        for e in edges.iter() {
            lo = lo.min(e[2]);
            hi = hi.max(e[2]);
        }

        let target = online.len() as i32 - 1;
        let mut q: VecDeque<i32> = VecDeque::new();
        let mut costs = vec![0_i64; online.len()];
        let adj = Self::calc_adj(edges, online);
        let mut deg = Self::calc_deg(&adj);

        if !Self::check(&adj, &mut q, &mut costs, &mut deg, lo, target, k) {
            return -1;
        }

        while lo <= hi {
            let mid = lo + (hi - lo) / 2;
            if Self::check(&adj, &mut q, &mut costs, &mut deg, mid, target, k) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        hi
    }
}
